File: hw1/python/main.py
# Main program for first homework of Utilization of datastructures
# and Tietorakenteet ja algoritmit.

import os
import random
import string
import timeit
import time
import gc
import sys
import binascii
import random

from datastructure import Datastructure

# Constants for commands
ADD = 'A'
READ = 'R'
PRINT = 'P'
MAXMIN = 'X'
EMPTY = 'E'
QUIT = 'Q'

GENERATE = 'G'
EFFICIENCY = 'F'
TEST = 'T'

# Other constants
OLD = 'old'
YOUNG = 'young'
E_FILE_NOT_FOUND = 'Error: file not found'
E_INT_EXPECTED = 'Error: birth year and enlisting year must be integers'
PROMPT = '> '
SEPARATOR = ' '

def test(datastructure, test_range=30):
    datastructure.empty()

    def test_print(datastructure):
        generate(datastructure, 1)
        datastructure.print()

    with open("sort.dat", 'w', encoding="utf-8") as output_file:
        output_file.write("#x y\n")

        for j in range(test_range):
            generate(datastructure, 250)
            t = timeit.Timer(lambda: test_print(datastructure), 'gc.enable()')
            sum_ = t.timeit(number=5) / 5
            size = datastructure.size()
            output_file.write("{0:d} {1:f}\n".format(size, sum_))
        datastructure.empty()

    def test_add(datastructure):
        generate(datastructure, 10)

    def test_oldest(datastructure):
        generate(datastructure,1)
        datastructure.oldest()

    def test_youngest(datastructure):
        generate(datastructure,1)
        datastructure.youngest()

    add_file = open('add.dat', 'w', encoding="utf-8")
    add_file.write("#x y\n")
    old_file = open('old.dat', 'w', encoding="utf-8")
    old_file.write("#x y\n")
    young_file = open('young.dat', 'w', encoding="utf-8")
    young_file.write("#x y\n")

#        t1 = timeit.Timer(lambda: test_add(datastructure), 'gc.enable()')
    t1 = timeit.Timer(lambda: generate(datastructure, 10), 'gc.enable()')
    t2 = timeit.Timer(lambda: test_oldest(datastructure), 'gc.enable()')
    t3 = timeit.Timer(lambda: test_youngest(datastructure), 'gc.enable()')
    for j in range(test_range):
        generate(datastructure, 250)
        add_sum = 0.0
        old_sum = 0.0
        young_sum = 0.0
        for i in range(5):
            add_sum += t1.timeit(number=1)
            old_sum += t2.timeit(number=1)
            young_sum += t3.timeit(number=1)
        size = datastructure.size()
        add_file.write("{0:d} {1:f}\n".format(size, add_sum / 5))
        old_file.write("{0:d} {1:f}\n".format(size, old_sum / 5))
        young_file.write("{0:d} {1:f}\n".format(size, young_sum / 5))
    add_file.close()
    old_file.close()
    young_file.close()

    datastructure.empty()

    del_file = open('del.dat', 'w', encoding="utf-8")
    del_file.write("#x y\n")
    t4 = timeit.Timer(lambda: datastructure.empty())
    for j in range(1, test_range + 1):
        generate(datastructure, j*250)
        size = datastructure.size()
        sum_ = t4.timeit(number=5) / 5
        del_file.write("{0:d} {1:f}\n".format(size, sum_))
    del_file.close()

    datastructure.empty()

def generate(datastructure, amount):
    colors = ["gold", "blue", "red"]

    for i in range(amount):
        # These are 2 times slower than method below, so use faster, but not
        # so pythonic methods to acquire random values for rank & name
#        rank = ''.join(random.choice(string.ascii_letters) for j in range(8))
#        name = ''.join(random.choice(string.ascii_letters) for j in range(10))
        rank = '%018x' % random.randrange(16**18)
        name = rank[8:]
        rank = rank[:8]

        year = random.randint(0, 200000)
        enlisting_year = random.randint(0,60) + 2200 + year
        birth_year = year + 2200
        shirt_color = random.choice(colors)
        datastructure.add(rank, birth_year, enlisting_year, shirt_color,
                          name)

def efficiency_test(datastructure, amount):
    # Generate first, because python is slow to generate random data
    generate(datastructure, amount)

    start = time.time()
    datastructure.print()
    datastructure.oldest()
    datastructure.youngest()
    datastructure.print()
    datastructure.print()
    datastructure.print()
    datastructure.print()
    datastructure.print()
    datastructure.print()
    datastructure.print()
    end = time.time()
    print("{0:f}".format(end - start), file=sys.stderr)

def read_file(filename, datastructure):
    """ Reads data from filename and adds it to the datastructure
    (READ command) """
    if not os.path.isfile(filename):
        print(E_FILE_NOT_FOUND)
        return

    with open(filename, 'r', encoding='utf-8') as input_file:
        for line in input_file:
            input_line = line.split(SEPARATOR)
            # Skip invalid lines
            if len(input_line) < 5:
                continue
            rank = input_line[0]
            try:
                birth_year = int(input_line[1])
                enlisting_year = int(input_line[2])
            except ValueError as err:
                print(E_INT_EXPECTED)
                return
            shirt_color = input_line[3]
            name = SEPARATOR.join(input_line[4:]).rstrip('\n')
            datastructure.add(rank, birth_year, enlisting_year, shirt_color,
                              name)


def main():
    """ Main program contains a simple shell which handles input and uses the
    datastructure accordingly """

    ds = Datastructure()
    input_line = ''
    command = ''
    while command != QUIT:
        
        try:
            input_line = input(PROMPT).split()
        except EOFError as err:
            print("")
            return

        # Skip empty lines
        if len(input_line) == 0:
            continue
        command = input_line[0]
        
        if command == ADD and len(input_line) >= 6:
            rank = input_line[1]
            try:
                birth_year = int(input_line[2])
                enlisting_year = int(input_line[3])
            except ValueError as err:
                print(E_INT_EXPECTED)
                continue

            shirt_color = input_line[4]
            name = SEPARATOR.join(input_line[5:])

            print(command, rank, birth_year, enlisting_year,
                  shirt_color, name, sep=SEPARATOR)
            
            ds.add(rank, birth_year, enlisting_year, shirt_color, name)
        elif command == READ and len(input_line) == 2:
            filename = input_line[1]
            print(command, filename, sep=SEPARATOR)
            read_file(filename, ds)
        elif command == PRINT:
            print(command, sep=SEPARATOR)
            ds.print()
        elif command == MAXMIN and len(input_line) == 2:
            parameter = input_line[1]
            print(command, parameter, sep=SEPARATOR)
            if parameter == OLD:
                ds.oldest()
            elif parameter == YOUNG:
                ds.youngest()
        elif command == EMPTY:
            print(command, sep=SEPARATOR)
            ds.empty()
        elif command == GENERATE and len(input_line) == 2:
            amount = int(input_line[1])
            generate(ds, amount)
        elif command == EFFICIENCY and len(input_line) == 2:
            amount = int(input_line[1])
            efficiency_test(ds, amount)

        elif command == TEST:
            if len(input_line) == 2:
                count = int(input_line[1])
            else:
                count = 30
            test(ds, count)
    print(command)

if __name__ == "__main__":
    try:
        main()
    except KeyboardInterrupt as err:
        print("")

File: hw1/cpp/main.cc
// Main program for first homework of Utilization of datastructures
// and Tietorakenteet ja algoritmit.

#include <iostream>
#include <cstdlib>
#include <string>
#include <fstream>
#include <iomanip>
#include <sstream>
#include <ctime>
#include <chrono>

#include "datastructure.hh"

// Known commands as constant chars
const char ADD = 'A';
const char READ = 'R';
const char PRINT = 'P';
const char MAXMIN = 'X';
const char EMPTY = 'E';
const char QUIT = 'Q';

const char GENERATE = 'G';
const char TEST = 'T';
const char EFFICIENCY = 'F';

// Constant MAX is used when asking biggest value and MIN for smallest
const std::string OLD = "old";
const std::string YOUNG = "young";

const std::string E_FILE_NOT_FOUND = "Error: file not found.\n";

const std::string PROMPT = "> ";
//defined in datastructure.hh
//const std::string SEPARATOR = " "



// Function readFile is used to read one input file at time.
void readFile(std::string const& filename, Datastructure& ds);

// Function generate is used to generate nodes to ds.
void generate(Datastructure& ds, unsigned int amount);

// Function efficiencyTest is used to test efficiency
void efficiencyTest(Datastructure& ds, unsigned int amount);

// Main program contains simple shell and it uses datastructure and
// handles error situations
int main() {

    std::srand(static_cast<unsigned int>(std::time(0)));
    char command = QUIT;
    std::string line;
    Datastructure ds;

    std::string rank;
    std::string name;
    unsigned int birthYear;
    unsigned int enlistingYear;
    std::string shirtColor;


    std::chrono::time_point<std::chrono::high_resolution_clock> start, end;


    // Simple shell reads command and it's parameters if any and then
    // executes the command
    do {
        std::cout << PROMPT;
        std::cin >> command;
        std::cout << command;

        if(command == ADD) {
            std::cin >> rank >> birthYear >> enlistingYear >> shirtColor;
            std::getline(std::cin, name);
            name = name.substr(1);
            std::cout << SEPARATOR << rank << SEPARATOR << birthYear << SEPARATOR
                << enlistingYear << SEPARATOR << shirtColor << SEPARATOR <<  name
                << std::endl;
            ds.add(rank, birthYear, enlistingYear, shirtColor, name);
        }
        if(command == READ) {
            std::string filename;
            std::cin >> filename;
            std::cout << SEPARATOR << filename << std::endl;
            readFile(filename, ds);
        }
        else if(command == PRINT) {
            std::cout << std::endl;
            ds.print();
        }

        else if (command == TEST)
        {
            ds.empty();
            std::ofstream file("sort.dat");
            file << "#x y" << std::endl;

            for ( int j = 0; j < 30; ++j )
            {
                double sum = 0.0;
                generate(ds, 250);
                for ( int i = 0; i < 5; ++i)
                {
                    start = std::chrono::system_clock::now();
                    generate(ds,1);
                    ds.print();
                    end = std::chrono::system_clock::now();
                    std::chrono::duration<double> elapsed = end-start;
                    sum += elapsed.count();
                }

                file << ds.size() << ' ' << sum/5.0 << std::endl;
            }
            file.close();
            ds.empty();

            std::ofstream addfile("add.dat");
            addfile << "#x y" << std::endl;
            std::ofstream oldfile("old.dat");
            oldfile << "#x y" << std::endl;
            std::ofstream youngfile("young.dat");
            youngfile << "#x y" << std::endl;

            for ( int j = 0; j < 30; ++j )
            {
                double addsum = 0.0;
                double oldsum = 0.0;
                double youngsum = 0.0;
                generate(ds, 250);
                for ( int i = 0; i < 5; ++i)
                {
                    start = std::chrono::system_clock::now();
                    generate(ds, 10);
                    end = std::chrono::system_clock::now();
                    std::chrono::duration<double> elapsed = end-start;
                    addsum += elapsed.count();

                    start = std::chrono::system_clock::now();
                    generate(ds,1);
                    ds.oldest();
                    end = std::chrono::system_clock::now();
                    elapsed = end-start;
                    oldsum += elapsed.count();

                    start = std::chrono::system_clock::now();
                    generate(ds,1);
                    ds.youngest();
                    end = std::chrono::system_clock::now();
                    elapsed = end-start;
                    youngsum += elapsed.count();

                }

                addfile << ds.size() << ' ' << addsum/5.0 << std::endl;
                oldfile << ds.size() << ' ' << oldsum/5.0 << std::endl;
                youngfile << ds.size() << ' ' << youngsum/5.0 << std::endl;

            }

            addfile.close();
            oldfile.close();
            youngfile.close();

            ds.empty();
            std::ofstream delfile("del.dat");
            delfile << "#x y" << std::endl;

            for ( unsigned int j = 1; j <= 30; ++j )
            {
                double sum = 0.0;
                generate(ds, j*250);
                size_t size = ds.size();
                for ( int i = 0; i < 5; ++i)
                {
                    start = std::chrono::system_clock::now();
                    ds.empty();
                    end = std::chrono::system_clock::now();
                    std::chrono::duration<double> elapsed = end-start;
                    sum += elapsed.count();
                }


                delfile << size << ' ' << sum/5.0 << std::endl;
            }

            delfile.close();
            ds.empty();
        }

        else if(command == MAXMIN) {
            std::string parameter;
            std::cin >> parameter;
            std::cout << SEPARATOR << parameter << std::endl;
            if(parameter == OLD) {
                ds.oldest();
            }
            else if(parameter == YOUNG) {
                ds.youngest();
            }
        }
        else if(command == EMPTY) {
            std::cout << std::endl;
            ds.empty();
        }
        else if (command == GENERATE)
        {
            unsigned int amount = 0;
            std::cin >> amount;
            //std::cout << SEPARATOR << "Generating " << amount << " nodes to datatructure" << std::endl;
            generate(ds, amount);

        }
        else if (command == EFFICIENCY)
        {
            unsigned int amount = 0;
            std::cin >> amount;
            //std::cout << SEPARATOR << "Generating " << amount << " nodes to datatructure" << std::endl;
            efficiencyTest(ds, amount);

        }

    } while(command != QUIT);

    std::cout << std::endl;


    return EXIT_SUCCESS;
}

// Function creates filestream from given filename and checks file
// existence. If file could not be read, prints error message. Then
// function reads all data from file and adds the data to the data
// structure.
void readFile(std::string const& filename, Datastructure& ds) {

    std::ifstream file(filename.c_str());

    if(!file) {
        file.close();
        std::cout << E_FILE_NOT_FOUND;
        return;
    }

    std::string line;
    std::string rank;
    std::string name;
    unsigned int birthYear;
    unsigned int enlistingYear;
    std::string shirtColor;


    // While loop reads file line by line and adds one ship at time to
    // the data structure.
    while(std::getline(file, line)) {

        std::istringstream linestream(line);
        linestream >> rank >> birthYear >> enlistingYear >> shirtColor;
        std::getline(linestream, name);
        name = name.substr(1);
        ds.add(rank, birthYear, enlistingYear, shirtColor, name);
    }

    file.close();
}


void generate(Datastructure& ds, unsigned int amount)
{
    static const char alphas[] = "abcdefghijklmnopqrstuvwxyz";

    for( unsigned int j = 0; j < amount; ++j )
    {
        std::string rank (8, ' ');
        for ( size_t i = 0; i < 8; ++i )
        {
            rank[i] = alphas[(unsigned int)std::rand() % (sizeof(alphas) - 1)];
        }

        std::string name (10, ' ');

        for ( size_t i = 0; i < 10; ++i )
        {
            name[i] = alphas[(unsigned int)std::rand() % (sizeof(alphas) - 1)];
        }

        unsigned int year = (unsigned int)std::rand() % 200000;
        unsigned int enlistingYear = (unsigned int)std::rand() % 60 + 2200 + year;
        unsigned int birthYear = year + 2200;

        static const std::string colors[3] = {"gold", "blue", "red" };

        std::string shirtColor = colors[(unsigned int)std::rand() % 3];



        ds.add(rank, birthYear, enlistingYear, shirtColor, name);
    }


}

void efficiencyTest(Datastructure& ds, unsigned int amount)
{
   generate(ds, amount);
   
   std::chrono::time_point<std::chrono::high_resolution_clock> start, end;
   start = std::chrono::system_clock::now();
   ds.print();
   ds.oldest();
   ds.youngest();
   ds.print();
   ds.print();
   ds.print();
   ds.print();
   ds.print();
   ds.print();
   ds.print();
   end = std::chrono::system_clock::now();
   std::chrono::duration<double> elapsed = end-start;
   std::cerr << elapsed.count() << std::endl;
}

// vim: set expandtab:tabstop=5:softtabstop=8:shiftwidth=8
